content/blog/elevate-your-v-project-with-unit-tests.md

```diff
@@ -0,0 +1,749 @@
+		size: 2
+		symbol: `🀃`
+	}
+	shape := geometry.generate_square(opts)
+	assert shape == ['🀃🀃', '🀃🀃']
+}
+
+// -----------------------------------------------------------------------------
+// Diamond
+// -----------------------------------------------------------------------------
+
+// Test to generate a diamond of size 5 without custom symbol
+fn test_generate_diamond_big_7() {
+	opts := geometry.ShapeOptions{
+		size: 5
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape == ['    *', '   ***', '  *****', ' *******', '*********', ' *******', '  *****',
+		'   ***', '    *']
+}
+
+// Test to generate a small diamond of size 2 without custom symbol
+fn test_generate_diamond_small_2() {
+	opts := geometry.ShapeOptions{
+		size: 2
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape == [' *', '***', ' *']
+}
+
+// Test to generate a small diamond of size 2 with a custom symbol
+fn test_generate_diamond_small_2_custom() {
+	opts := geometry.ShapeOptions{
+		size: 2
+		symbol: `x`
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape == [' x', 'xxx', ' x']
+}
+
+// Invalid diamond
+fn test_generate_diamond_invalid() {
+	opts := geometry.ShapeOptions{
+		size: -1
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape.len == 0
+}
+
+fn test_generate_diamond_unicode_1() {
+	opts := geometry.ShapeOptions{
+		size: 3
+		symbol: `🔼`
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape == ['  🔼', ' 🔼🔼🔼', '🔼🔼🔼🔼🔼', ' 🔼🔼🔼', '  🔼']
+}
+
+fn test_generate_diamond_unicode_2() {
+	opts := geometry.ShapeOptions{
+		size: 2
+		symbol: `শ`
+	}
+	shape := geometry.generate_diamond(opts)
+	assert shape == [' শ', 'শশশ', ' শ']
+}
+
+```
+
+The source for `quadrilateral_test.v` is available
+[here](https://github.com/hungrybluedev/geo/blob/main/test/unit_tests/quadrilateral_test.v).
+
+### Observations
+
+You might have noticed that we're using unicode characters in the tests.
+We've made our implementation robust enough to support them.
+
+You can add more tests if you want. We do not need to consider all the
+possibilities cases. It depends on how they are prioritised. Ideally,
+every single public facing function should have an accompanying unit
+test, and the test suite must execute all of the relevant lines in the
+project (which we refer to as _coverage_).
+
+Practically, we settle for certain thresholds (like 90% coverage), or unit
+testing only the critical, unchanging components. If you're in a team, listen
+to what your senior says. If you're the senior, don't set unrealistic
+expectations for your team, or yourself.
+
+## BONUS: Testing The Output of External Commands
+
+Until now, we assessed the correctness of our implementation my testing
+the functions of the `geometry` module directly. We would also like to
+make sure the program can compile and generate coherent output. In order
+to test this, we can spawn a new process from V and compare the output
+of the child process to our expected results.
+
```

Comment:
```suggestion
make sure the program can compile and generate coherent output. In order
to test this, we can make our test, compile and run a separate V program, and
compare its output to our expected results.

```

Reply:
```suggestion
make sure the program can compile and generate coherent output. In order
to test this, we can make our test compile and run a separate V program, and
compare its output to our expected results.

```
Of course, this is functional testing, rather than unit testing.

Reply:
yes